### PR TITLE
Fix erasure of this types and refactor value class erasure

### LINKED ISSUE #682: unittest_collection.scala: java.lang.NoSuchMethodError scala.collection.mutable.BufferLike.$plus$eq(Ljava/lang/Object;)Lscala/collection/mutable/BufferLike;
```
[info] % dotc unittest_collection.scala
[info] % java \
[info]   -classpath \
[info]   /Users/dark/workspace/dotty/tests/partest-generated/run/unittest_collection-run.obj:/Users/dark/workspace/dotty/target/scala-2.11/dotty_2.11-0.1-SNAPSHOT.jar:/Users/dark/.ivy2/cache/me.d-d/scala-compiler/jars/scala-compiler-2.11.5-20150619-173733-3bcd390afa.jar:/Users/dark/.ivy2/cache/org.scala-lang/scala-reflect/jars/scala-reflect-2.11.5.jar:/Users/dark/.ivy2/cache/org.scala-lang/scala-library/jars/scala-library-2.11.5.jar \
[info]   -Dfile.encoding=UTF-8 \
[info]   -Djava.library.path=/Users/dark/workspace/dotty/tests/partest-generated/run \
[info]   -Dpartest.output=/Users/dark/workspace/dotty/tests/partest-generated/run/unittest_collection-run.obj \
[info]   -Dpartest.lib=/Users/dark/.ivy2/cache/org.scala-lang/scala-library/jars/scala-library-2.11.5.jar \
[info]   -Dpartest.reflect=/Users/dark/.ivy2/cache/org.scala-lang/scala-reflect/jars/scala-reflect-2.11.5.jar \
[info]   -Dpartest.comp=/Users/dark/workspace/dotty/target/scala-2.11/dotty_2.11-0.1-SNAPSHOT.jar \
[info]   -Dpartest.cwd=/Users/dark/workspace/dotty/tests/partest-generated/run \
[info]   -Dpartest.test-path=/Users/dark/workspace/dotty/tests/partest-generated/run/unittest_collection.scala \
[info]   -Dpartest.testname=unittest_collection \
[info]   -Djavacmd=java \
[info]   -Djavaccmd=javac \
[info]   -Duser.language=en \
[info]   -Duser.country=US \
[info]   scala.tools.nsc.MainGenericRunner \
[info]   -usejavacp \
[info]   Test \
[info]   jvm > unittest_collection-run.log
[info] java.lang.NoSuchMethodError: scala.collection.mutable.BufferLike.$plus$eq(Ljava/lang/Object;)Lscala/collection/mutable/BufferLike;
[info] 	at Test$.test(unittest_collection.scala:21)
[info] 	at Test$.main(unittest_collection.scala:11)
[info] 	at Test.main(unittest_collection.scala)
[info] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[info] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[info] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[info] 	at java.lang.reflect.Method.invoke(Method.java:497)
[info] 	at scala.reflect.internal.util.ScalaClassLoader$$anonfun$run$1.apply(ScalaClassLoader.scala:70)
[info] 	at scala.reflect.internal.util.ScalaClassLoader$class.asContext(ScalaClassLoader.scala:31)
[info] 	at scala.reflect.internal.util.ScalaClassLoader$URLClassLoader.asContext(ScalaClassLoader.scala:101)
[info] 	at scala.reflect.internal.util.ScalaClassLoader$class.run(ScalaClassLoader.scala:70)
[info] 	at scala.reflect.internal.util.ScalaClassLoader$URLClassLoader.run(ScalaClassLoader.scala:101)
[info] 	at scala.tools.nsc.CommonRunner$class.run(ObjectRunner.scala:22)
[info] 	at scala.tools.nsc.ObjectRunner$.run(ObjectRunner.scala:39)
[info] 	at scala.tools.nsc.CommonRunner$class.runAndCatch(ObjectRunner.scala:29)
[info] 	at scala.tools.nsc.ObjectRunner$.runAndCatch(ObjectRunner.scala:39)
[info] 	at scala.tools.nsc.MainGenericRunner.runTarget$1(MainGenericRunner.scala:65)
[info] 	at scala.tools.nsc.MainGenericRunner.run$1(MainGenericRunner.scala:87)
[info] 	at scala.tools.nsc.MainGenericRunner.process(MainGenericRunner.scala:98)
[info] 	at scala.tools.nsc.MainGenericRunner$.main(MainGenericRunner.scala:103)
[info] 	at scala.tools.nsc.MainGenericRunner.main(MainGenericRunner.scala)
```

### PR DESCRIPTION
**Pull request by odersky**
Thistypes erased to the underlying class. This is wrong. When seen as part of some other type,
a ThisType has to erase to the erasure of the underlying type (i.e. the erasure if the selftype
of the class). unittest-collections.scala failed with a MethodNotFound error because the erasure
was computed incorrectly.

On the other hand, a tree with a ThisType type, should keep the type, analogous to a
tree with a TermRef type.

Also, refactored value class erasure to avoid spreading semiEraseVCs.

Fixes #682 

Rview by @DarkDimius @smarter

### CODE REVIEW

**Comment by odersky** [comment]
rebased to master

File: src/dotty/tools/dotc/transform/ExtensionMethods.scala [hunk 1]
```diff
@@ -14,7 +14,7 @@ import core._
 import Phases.Phase
 import Types._, Contexts._, Constants._, Names._, NameOps._, Flags._, DenotTransformers._
 import SymDenotations._, Symbols._, StdNames._, Annotations._, Trees._, Scopes._, Denotations._
-import TypeErasure.{ erasure, ErasedValueType }
+import TypeErasure.{ erasure, valueErasure, ErasedValueType }
```
Comment (smarter) [inline src/dotty/tools/dotc/transform/ExtensionMethods.scala]: `erasure` does not need to be imported anymore.

**Comment by smarter** [comment]
LGTM otherwise, it's nice that this simplified erasure a bit.

**Comment by odersky** [comment]
@smarter I dropped the import in #699

**Pull request merged**